### ISSUE TITLE
postgres not working

### ISSUE THREAD

**Issue by happyshows** [rstudio/pool#37]
Hi, 

I was trying the following example you listed in another issue. Below are the code & error message.

```
library(dplyr)
library(pool)
library(DBI)
library(RPostgreSQL)

pool <- dbPool(
  drv = RPostgreSQL::PostgreSQL(max.con=20),
  dbname = "postgresdemo",
  host = "postgresdemo.cfd8mtk93q6a.us-west-2.rds.amazonaws.com",
  user = "guest",
  password = "guest"
)
src_pool(pool) %>% tbl("city") %>% head(n + 1) 
```

```
Error in UseMethod("tbl") : 
  no applicable method for 'tbl' applied to an object of class "c('src_postgres', 'src_sql', 'src')"
```

**Comment by bborgesr** [comment]
Hm, it works fine for me with the following:

```r
> devtools::session_info()
Session info -------------------------------------------------------------------------------
 setting  value                                 
 version  R version 3.3.2 RC (2016-10-26 r71594)
 system   x86_64, darwin13.4.0                  
 ui       RStudio (1.1.107)                     
 language (EN)                                  
 collate  en_US.UTF-8                           
 tz       Europe/Lisbon                         
 date     2017-03-23                            

Packages -----------------------------------------------------------------------------------
 package     * version    date       source                       
 assertthat    0.1        2013-12-06 CRAN (R 3.3.0)               
 DBI         * 0.6        2017-03-09 cran (@0.6)                  
 devtools      1.12.0     2016-06-24 CRAN (R 3.3.0)               
 digest        0.6.12     2017-01-27 CRAN (R 3.3.1)               
 dplyr       * 0.5.0      2016-06-24 CRAN (R 3.3.1)               
 htmltools     0.3.5      2016-03-21 CRAN (R 3.3.0)               
 httpuv        1.3.3      2015-08-04 CRAN (R 3.3.1)               
 magrittr      1.5        2014-11-22 CRAN (R 3.3.1)               
 memoise       1.0.0      2016-01-29 CRAN (R 3.3.1)               
 mime          0.5        2016-07-07 CRAN (R 3.3.1)               
 pool        * 0.1.0      2017-03-23 Github (rstudio/pool@52199b0)
 R6            2.2.0      2016-10-05 CRAN (R 3.3.1)               
 Rcpp          0.12.9     2017-01-14 CRAN (R 3.3.1)               
 RPostgreSQL * 0.4-1      2016-05-08 CRAN (R 3.3.0)               
 rstudioapi    0.6        2016-06-27 CRAN (R 3.3.1)               
 shiny         1.0.0.9001 2017-03-22 local                        
 tibble        1.2        2016-08-26 CRAN (R 3.3.0)               
 withr         1.0.2      2016-06-20 CRAN (R 3.3.1)               
 xtable        1.8-2      2016-02-05 CRAN (R 3.3.1)               
 yaml          2.1.14     2016-11-12 CRAN (R 3.3.2)  
```

What package versions are you using?

**Comment by happyshows** [comment]
```
> devtools::session_info()
Session info ------------------------------------------
 setting  value                       
 version  R version 3.3.1 (2016-06-21)
 system   x86_64, darwin13.4.0        
 ui       RStudio (1.0.136)           
 language (EN)                        
 collate  en_US.UTF-8                 
 tz       America/Chicago             
 date     2017-03-23                  

Packages ----------------------------------------------
 package     * version    date      
 assertthat    0.1        2013-12-06
 DBI         * 0.6        2017-03-23
 devtools      1.12.0     2016-06-24
 digest        0.6.12     2017-01-27
 dplyr       * 0.5.0.9000 2017-03-05
 magrittr      1.5        2014-11-22
 memoise       1.0.0      2016-01-29
 pool        * 0.1.0      2017-03-23
 R6            2.2.0      2016-10-05
 Rcpp          0.12.10    2017-03-23
 RPostgreSQL * 0.4-1      2016-05-08
 tibble        1.2        2016-08-26
 withr         1.0.2      2016-06-20
 source                        
 CRAN (R 3.3.0)                
 Github (rstats-db/DBI@76d8d65)
 CRAN (R 3.3.0)                
 cran (@0.6.12)                
 Github (hadley/dplyr@f39db50) 
 CRAN (R 3.3.0)                
 CRAN (R 3.3.0)                
 Github (rstudio/pool@52199b0) 
 CRAN (R 3.3.0)                
 Github (RcppCore/Rcpp@5a71cf6)
 cran (@0.4-1)                 
 CRAN (R 3.3.0)                
 CRAN (R 3.3.0)                
```

**Comment by bborgesr** [comment]
try updating `dplyr` to the devel version and see if it fixes it? This new version of `dplyr` should be released to CRAN soon and when I start working on `pool` again, that will be my baseline (Hadley is changing up a few things related to DBs in the new `dplyr` version).

**Comment by happyshows** [comment]
same error after installation of the dev dplyr. Can you confirm whether you're on cran version or latest dev?

**Comment by bborgesr** [comment]
Which package ware you talking about? I also tried with CRAN versions of everything (expect `pool`) and it worked fine. This was yesterday. Today, for some reason, it seems like the RDS db is down at the moment, because I always get a "connection refused" error (but that also happens from the terminal, so it's not an R issue).

**Comment by happyshows** [comment]
I'm referring to dplyr. It seems you're using CRAN version of dplyr based on sessionInfo()

```
dplyr       * 0.5.0      2016-06-24 CRAN (R 3.3.1)     
```

**Comment by bborgesr** [comment]
That's correct. Sorry, my suggestion before to update to the devel version was silly. Since Hadley is doing the changes I mentioned, pool is not yet necessarily compatible with that. Once that ships to CRAN, I'll definitely make sure that `pool` is updated to match (some of these changes were actually requested by us, so that the integration between pool and dplyr becomes better). But for now that isn't the case.

This is an even easier test (no remote DBs):

```r
library(dplyr)
library(pool)
library(DBI)
library(RSQLite)

pool <- dbPool(RSQLite::SQLite(), ":memory:")
dbWriteTable(pool, "cars", head(cars, 15))
src_pool(pool) %>% tbl("cars") %>% head(5) 
```

I've just confirmed that this works fine in CRAN dplyr, but not devel dplyr.

Hopefully, dplyr will be released soon and we'll adapt (for now, I'm keeping it compatible with CRAN only)

**Comment by happyshows** [comment]
cool, thanks for the update.

I'll leave this issue open for now. Hope this is fine with you.

**Comment by happyshows** [comment]
just ran the two examples. The one using SQLite works fine. But the remote postgres ones throw out warning messages:

```
> src_pool(pool) %>% tbl("city") %>% head(3) 
Source:   query [?? x 5]
Database: postgres 9.5.2 [guest@postgresdemo.cfd8mtk93q6a.us-west-2.rds.amazonaws.com:5432/postgresdemo]

     id     name countrycode district population
  <int>    <chr>       <chr>    <chr>      <int>
1     1    Kabul         AFG    Kabol    1780000
2     2 Qandahar         AFG Qandahar     237500
3     3    Herat         AFG    Herat     186800
Warning messages:
1: You still have checked out objects. Return them to the pool so they can be destroyed. (If these are leaked objects - no reference - they will be destroyed the next time the garbage collector runs). 
2: In force(expr) : You have a leaked pooled object. Destroying it.
```

Is this expected? In the code, there's no poolCheckout, why it's still defaulting to the DBIConnection approach?

**Comment by bborgesr** [comment]
That is exactly part of why we needed dplyr to change - for now the integration is pretty flimsy, but I'm still surprised those warnings get printed... I'll look into this when I start the integration from the `pool` side. Sorry I can't give you a better answer at this time...

**Comment by bborgesr** [comment]
Also, I don't get those errors on a clean session:

```r
Restarting R session...

> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> library(pool)
> library(DBI)
> library(RPostgreSQL)
> 
> pool <- dbPool(
+     drv = RPostgreSQL::PostgreSQL(max.con=20),
+     dbname = "postgresdemo",
+     host = "postgresdemo.cfd8mtk93q6a.us-west-2.rds.amazonaws.com",
+     user = "guest",
+     password = "guest"
+ )
> src_pool(pool) %>% tbl("city") %>% head(6) 
Source:   query [?? x 5]
Database: postgres 9.5.2 [guest@postgresdemo.cfd8mtk93q6a.us-west-2.rds.amazonaws.com:5432/postgresdemo]

     id           name countrycode      district population
  <int>          <chr>       <chr>         <chr>      <int>
1     1          Kabul         AFG         Kabol    1780000
2     2       Qandahar         AFG      Qandahar     237500
3     3          Herat         AFG         Herat     186800
4     4 Mazar-e-Sharif         AFG         Balkh     127800
5     5      Amsterdam         NLD Noord-Holland     731200
6     6      Rotterdam         NLD  Zuid-Holland     593321

> devtools::session_info()
Session info ----------------------------------------------------------------------------------------------
 setting  value                                 
 version  R version 3.3.2 RC (2016-10-26 r71594)
 system   x86_64, darwin13.4.0                  
 ui       RStudio (1.1.122)                     
 language (EN)                                  
 collate  en_US.UTF-8                           
 tz       Europe/Lisbon                         
 date     2017-03-24                            

Packages --------------------------------------------------------------------------------------------------
 package     * version date       source                       
 assertthat    0.1     2013-12-06 CRAN (R 3.3.0)               
 DBI         * 0.6     2017-03-09 CRAN (R 3.3.2)               
 devtools      1.12.0  2016-06-24 CRAN (R 3.3.0)               
 digest        0.6.12  2017-01-27 CRAN (R 3.3.1)               
 dplyr       * 0.5.0   2016-06-24 CRAN (R 3.3.0)               
 magrittr      1.5     2014-11-22 CRAN (R 3.3.1)               
 memoise       1.0.0   2016-01-29 CRAN (R 3.3.1)               
 pool        * 0.1.0   2017-03-23 Github (rstudio/pool@52199b0)
 R6            2.2.0   2016-10-05 CRAN (R 3.3.1)               
 Rcpp          0.12.10 2017-03-19 cran (@0.12.10)              
 RPostgreSQL * 0.4-1   2016-05-08 CRAN (R 3.3.0)               
 tibble        1.2     2016-08-26 CRAN (R 3.3.0)               
 withr         1.0.2   2016-06-20 CRAN (R 3.3.1)               
 yaml          2.1.14  2016-11-12 CRAN (R 3.3.2)               
```

**Comment by happyshows** [comment]
One more thing, not sure this should be filed separately.

I put the following chunk in global.R and observe the status of postgres on linux box.
```
pool <- dbPool(RPostgres::Postgres(),dbname = 'comment', 
               host = 'XX.XX.XX.XX', 
               port = 5432, 
               user = 'shiny',
               password = '****')
```

each time when I launch a new shiny app instance, the connection (created in global.R) won't close even after shiny app has ended. I ran & closed 3 times of the shiny app which generates 19744, 19750 & 19833 idle threads. 
During the app session, there's no more thread being created though, so it looks fine.
Is there any reason why the pool object & its connection are not destroyed after shiny app terminates?

```
           ├─19614 postgres: shiny comment XX.XX.XX.XX(63060) idle
           ├─19619 postgres: shiny comment XX.XX.XX.XX(63065) idle
           ├─19720 postgres: shiny comment XX.XX.XX.XX(63263) idle
           ├─19744 postgres: shiny comment XX.XX.XX.XX(57418) idle
           ├─19750 postgres: shiny comment XX.XX.XX.XX(57436) idle
           └─19833 postgres: shiny comment XX.XX.XX.XX(57564) idle
```

**Comment by happyshows** [comment]
Sure, #38 created.

**Comment by jcheng5** [comment]
@wch This might be `SIGTERM` not running finalizers. I wonder if we should start thinking about installing signal handlers (or asking R-core to give us an officially sanctioned way to do this).

**Comment by wch** [comment]
@jcheng5 Sounds like a good idea. Will that work on Windows? If not, an alternative is to use the supervisor stuff I've been working on.

**Comment by jcheng5** [comment]
@happyshows What is that a screenshot of (the idle threads)? Is that from the server or the client?

**Comment by jcheng5** [comment]
Sorry, I mean the shiny server or the DB server. And if they're the same server, then the shiny-server process or the postgres process.

**Comment by happyshows** [comment]
hi Joe, it's from the postgres server. I was using Rstudio on my mac to test connections. No shiny server is involved.

**Comment by jcheng5** [comment]
So when an R process with active Postgres connections is killed, the Postgres server doesn't immediately realize it? Are the orphaned connections cleaned up on a schedule or heartbeat of something, or will they stay open forever?

**Comment by jcheng5** [comment]
Oh, oh, sorry, you're not actually ending the R process, are you? Just launching the Shiny app several times in the same process, and each time is leaking a connection?

**Comment by happyshows** [comment]
Launch app -> new connection established -> close app -> connection does not destroy... Launch app again -> new connection 2 established .. etc.   However, postgres (without bouncer) seem to be able to close idea connections after a few minutes.
I think it's leaking. Any further step I can take to debug this?

Currently I'm switching to pgbouncer to manage the leak.

**Comment by jcheng5** [comment]
Thanks, that's really helpful. @bborgesr and I discussed it, she's going to suggest some workarounds for the meantime but we will be sure to investigate it next week and figure out a good solution.

**Comment by bborgesr** [comment]
While we try to repo this, here's one task and one possible workaround for now:

### Ask
We suspect the problem is that `poolClose()` is not called explicitly and for some reason, the pool isn't getting garbage collected after the shiny app is closed (which is what should be happening). So, could you try doing `poolClose` (through an action button or something) before you close your app and let us know if that made the problem disappear?

### Workaround
If we're at all right in what may be happening here, the workaround for now is:

1. either always call `poolClose()` (if that worked)  once you're done with `pool` (which might be difficult/impossible for certain apps, because you may always need a pool available)

2. create the Pool object in the global environment and then check for its existence inside the app and only create a pool if you can't find it. So, in effect, you'd always have the same pool, so you shouldn't experience the problem either.

I realize both workarounds are not great (and it may not even fix your problem), but hopefully it will be enough to tide you over as we figure it out.

**Comment by bborgesr** [comment]
The original issue in this thread had to do with dplyr compatibility and was solved in #42. The other issue has been filed separately, so we can close this one.

**Comment by happyshows** [comment]
@bborgesr can you provide a new hello world example for me to try with the new dev package?
